AGENTS.md

```diff
@@ -0,0 +1,254 @@
+
+### Development
+```bash
+# Build the project
+cargo build
+
+# Run all tests
+cargo test
+
+# Run a specific test
+cargo test test_name
+
+# Format code (required before committing)
+cargo fmt
+
+# Lint code
+cargo clippy
+
+# Check code without building
+cargo check
+```
+
+### CLI Usage (after building)
+```bash
+# Core commands
+pks check         # Look for violations
+pks validate      # Look for validation errors
+pks update        # Update package_todo.yml files
+pks create        # Create new packs
+```
+
+## Architecture Overview
+
+The codebase is organized around a modular checker system that validates various aspects of package boundaries:
+
+- **Entry Points**: `src/main.rs` (CLI) and `src/lib.rs` (library)
+- **Core Module**: `src/packs/` contains all business logic
+- **Error Handling**: Uses `anyhow` for simplified `Result` handling throughout the codebase
+- **Checker System**: Located in `src/packs/checker/`, implements different violation types:
+  - `dependency.rs` - Package dependency violations
+  - `privacy.rs` - Public API violations
+  - `visibility.rs` - Package visibility violations
+  - `layer.rs` - Architecture layer violations
+  - `folder_privacy.rs` - Folder-level privacy violations
+
+- **Parsing**: `src/packs/parsing/` handles Ruby and ERB code parsing using `lib-ruby-parser`
+- **Performance**: Uses `rayon` for parallel processing and implements caching in `src/packs/caching/`
+- **Configuration**: Managed through `src/packs/configuration.rs`, reads `packwerk.yml` and `package.yml` files
+
+## Testing
+
+Tests are fixture-based and located in `tests/`. Key test fixtures in `tests/fixtures/` represent different application scenarios. The test suite uses `assert_cmd` for CLI testing.
+
+To run tests for a specific checker or feature, use:
+```bash
+cargo test checker_name
+```
+
+## Key Implementation Details
+
+- **Rust version**: specified in `rust-toolchain.toml`
+- **Parallel Processing**: Heavily uses `rayon` for performance
+- **Serialization**: Uses `serde_yaml` for YAML files, with deterministic serialization for `package_todo.yml`
+- **Ruby Parsing**: Uses `lib-ruby-parser` to generate ASTs for Ruby code analysis
+- **Graph Operations**: Uses `petgraph` for dependency cycle detection
+
+## Important Files to Know
+
+- `src/packs/pack.rs` - Core Pack struct and package representation
+- `src/packs/package_todo.rs` - Manages TODO violations in `package_todo.yml` files
+- `src/packs/checker.rs` - Main violation checking orchestration
+- `src/packs/cli.rs` - CLI command definitions and argument parsing
+
+## Rust Style Guide
+
+This codebase follows idiomatic Rust patterns. Here are key conventions:
+
+### Error Handling
+
+This codebase follows a principled approach to error handling: **panic for programming bugs, Result for recoverable errors**.
+
+#### When to Use `.expect()` or `.unwrap()` (Panic)
+Use panics for programming bugs - situations that should never happen if the code is written correctly:
```

Comment:
I've been avoiding `panic` and instead propagating errors with `Result` return types. I could see the `packs` gem wrapping `pks` in the future. It would be a benefit for clients to decide how they want to handle errors.